core/rio/binary/src/main/java/org/eclipse/rdf4j/rio/binary/BinaryRDFWriterV1.java

```diff
@@ -0,0 +1,307 @@
+/*******************************************************************************
+ * Copyright (c) 2015 Eclipse RDF4J contributors, Aduna, and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.rio.binary;
+
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.BNODE_VALUE;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.COMMENT;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.DATATYPE_LITERAL_VALUE;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.END_OF_DATA;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.FORMAT_V1;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.LANG_LITERAL_VALUE;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.MAGIC_NUMBER;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.NAMESPACE_DECL;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.NULL_VALUE;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.STATEMENT;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.TRIPLE_VALUE;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.URI_VALUE;
+import static org.eclipse.rdf4j.rio.binary.BinaryRDFConstants.VALUE_REF;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.rdf4j.common.io.ByteSink;
+import org.eclipse.rdf4j.model.BNode;
+import org.eclipse.rdf4j.model.IRI;
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.Statement;
+import org.eclipse.rdf4j.model.Triple;
+import org.eclipse.rdf4j.model.Value;
+import org.eclipse.rdf4j.model.util.Literals;
+import org.eclipse.rdf4j.rio.RDFFormat;
+import org.eclipse.rdf4j.rio.RDFHandlerException;
+import org.eclipse.rdf4j.rio.RDFWriter;
+import org.eclipse.rdf4j.rio.helpers.AbstractRDFWriter;
+
+/**
+ * @author Arjohn Kampman
+ */
+public class BinaryRDFWriterV1 extends AbstractRDFWriter implements RDFWriter, ByteSink {
```

Comment:
Actually, looking back; I don't really like the redundancy in the code. I was a bit hesitant to change the V1 code, but I'll rewrite this. I'll introduce a `WriterSetting` for the format.

Reply:
While I'm at it, does it make sense to add `WriterSetting`s for other aspects as well? E.g. the buffer size, the charset and whether or not to recycle identifiers? Let me know what you think.

Reply:
I think for buffer size, format version, and recycling that makes sense. These options should probably all be in a new class `BinaryRDFWriterSettings`, and follow the same pattern we use in, for example, `BasicWriterSettings`. 

For charset I'm not sure. I'm not really seeing any benefit to making that configurable separately from the chosen format version. It would also make it harder to support compatibility, because if you make this configurable separately you'd have to add the chosen charset to the file header for the parser to figure it out (whereas now it just knows it based on the version). I suggest for now to just keep the charset strictly tied to the format version. V1 = UTF-16, V2 = UTF-8.

Reply:
I've added such settings [in the pull request](https://github.com/eclipse/rdf4j/pull/3052/files#diff-d6155f1a1e71c6dafe1a7108a5e61e4d656720fe36ce93b2c5f8e11870f80ec3R20).

My thinking with regards to the character set was that while I did some testing, I may have neglected a big part of the planet that is using characters outside of the ascii range. UTF-8 _can_ consume _more_ space.

E.g. the sentence 这是一个例子句子翻译成中国 consumes 39 bytes in UTF-8 and 26 bytes in UTF-16. Similar difference hold for e.g. Japanse, Birmese, Hindi and probably most Asian languages. While I think defaulting to UTF-8 is defensible as a lot of bytes will be consumed typically by IRIs and I haven't found any ontologies in e.g. Chinese just yet, I would like to give users the ability to use a string encoding that's suited for their needs.

In any case, it's not a massive burden to maintain. The cost is encoding the character set in the header of the binary RDF file (see [here](https://github.com/eclipse/rdf4j/pull/3052/files#diff-aa747cbb28c73e81e9da469f6c35ac098a18c361278988d814c3f1269edaf52fR124)).